### ISSUE TITLE
Add information about recurring event and the sequence number

### ISSUE THREAD

**Issue by NicoHood** [niccokunzmann/python-recurring-ical-events#90]
Hi!
I am using your library with success for some time. I do have a feature request, or more a brainstorm on how to solve this issue.

With your library I am basically unflattening the recurring event to flat single-day events. However the information that this **was** a recurring event is actually lost then. What is also lost is the information which recurring sequence this event was (the first, seconds, third occurrence).

Why do I need this? a) I have "normal" and recurring events. Sometimes I want to filter out the recurring events, as they bloat my usecase. b) I need a unique ID for each event. Usually the events have this id, but recurring events all have the same id. So I need to calculate a new internal id from the id + the sequence numer. That's at leats the plan.

Now how to solve this? Currently when parsing I will look at the whole input ical file and check if that original event had an rrule set. This way I can determine if the event was recurring. I currently cannot determine the sequence number, I'd need an rrule tool for that, I guess.

But isnt't there a way to built this into the library? Maybe with a custom `X-RECURRING` property or so? I don't want to bloat the library with my special usecase, but I can imagine, that others also might run into this issue. What are your thoughts?

**Comment by NicoHood** [comment]
I successfully managed to calculate the sequence as well. So here is my code, so you can have a better feeling of what I am talking about:

```py
        ical = icalendar.Calendar.from_ical(response.text)

# A dict that checks if a given uid is a recurring event.
# This is required since we unfold all events and afterwards cannot tell if they are recurring or not.
uid_recurring_map = {}
for component in ical.walk():
if component.name == "VEVENT":
    uid_recurring_map[component["UID"]] = component

# Get start and end date of week:
# Requires python 3.8: https://docs.python.org/3/library/datetime.html#datetime.date.fromisocalendar
start = date.fromisocalendar(year, week, 1)
end = start + timedelta(days=7)

events = recurring_ical_events.of(ical).between(start, end)

# Convert ical to our internal class formal
eventsData = list()
for event in events:
parent_event = uid_recurring_map[event["UID"]]
recurring = parent_event.get('RRULE') is not None
sequence = 0
if recurring:
    # Recreate the rrule using rrule, exdate, rdate.
    # Note: Exrule is deprecated and thereby not used.
    event_rrule = rrulestr(parent_event.get('RRULE').to_ical().decode(), dtstart=parent_event["DTSTART"].dt)

    # Rrules are managed in an rruleset:
    # https://dateutil.readthedocs.io/en/stable/rrule.html#dateutil.rrule.rruleset
    set = rruleset()
    set.rrule(event_rrule)

    # TODO do we even want to include the exdate and rdate into the sequence?
    # TODO maybe it would rearrange the whole sequence if one event is for example deleted.

    # This is a bit complicated:
    # EXDATE and be either an vDDDLists object or a list of vDDDLists objects.
    all_exdates = parent_event.get('EXDATE')
    if all_exdates is not None:
        # If it is not a list, simply create a list with that single entry for easier parsing.
        if not isinstance(all_exdates, list):
            all_exdates = [all_exdates]

        # Now iterate through the list, and the nested elements (also in a list)
        for exdatelist in all_exdates:
            for exdate in exdatelist.dts:
                set.exdate(exdate.dt)

    # This is a bit complicated:
    # RDATE and be either an vDDDLists object or a list of vDDDLists objects.
    all_rdates = parent_event.get('RDATE')
    if all_rdates is not None:
        # If it is not a list, simply create a list with that single entry for easier parsing.
        if not isinstance(all_rdates, list):
            all_rdates = [all_rdates]

        # Now iterate through the list, and the nested elements (also in a list)
        for rdatelist in all_rdates:
            for rdate in rdatelist.dts:
                set.rdate(rdate.dt)

    # We must expand the whole rrule in order to know which sequence number was used.
    # The sequence starts with 1 (not zero!)
    sequence = len(set.between(parent_event["DTSTART"].dt, event["DTSTART"].dt, inc=True))

data = self.get_event_data(event, recurring = recurring, sequence = sequence)
if data is not None:
    eventsData.append(data)

return eventsData
```

And this is where I generate the guids:
```py
# Generate a unique (but well-known) id for the event, by using the sequence of the recurring event.
guid = uuid.uuid5(uuid.NAMESPACE_URL, f"{event['UID']}-{sequence}")
```

**Comment by niccokunzmann** [comment]
Hi, thanks for taking the time for  the whole motivation of your use-case.

When I implemented this, I also needed to find out if a recurrence has a modification, so  there is identity already specified by the RFC 5545:
An event is the same event as another event if and only if these apply:
(1) the UUID is the same
(2) the DTSTART is the same

Or else: if the same event happens twice at the exact same time, it is the same. (If you make a copy in your calendar, the UUID changes)

If the sequence number is changed, we know which version is latest.

So, it seems, that the library just looses this information. That should be easy to add.

On a side-note, I would encourage you to use the identity specified by the standard and not construct your own for it like [event no 10 with UUID X]

I would say that knowing event identity is a common use-case, maybe the other issue #29 about the event index is about that - but without the motivation that is hard to guess. It should at least be documented in a section. I would say that this holds true and we can make a test with all calendar files used by this library.

```
counts = {} # event identity : count
for event in recurring_ical_events.of(a_calendar).all():
    event_identity = (event["UUID"], event["DTSTART"]) # assume UUID given
    counts[event_identity] = counts.get(event_identity, 0) + 1
assert all(count == 1 for count in counts.values())
```
What do you think?

**Comment by NicoHood** [comment]
> Numbering fails if a weekly event is moved by more than one week.

That is not true, why should it be? The recurrence id is set to the date where the event should originally take place. I take that into account and search for this end date, rather than the new, moved end date. This way I get the sequence number in the right place. What you mean is possibly that the order is wrong (correct), but the number itself is still unique (which is most important for me).

My usecase is quite specific. Another idea would be to integrate my logic and use this to produce unique UIDs. Because this is the final usecase, and if your library wants to unflatten an ical file, shouldn't it also take the UIDs into account? I could imagine that recurring events unflattend will have duplicated, UIDs which is agains the spec, isn't it?

**Comment by niccokunzmann** [comment]
> This way I get the sequence number in the right place.

Cool! You might have said that somewhere. I am catching up.

> I need a unique ID for each event. Usually the events have this id, but recurring events all have the same id. So I need to calculate a new internal id from the id + the sequence numer.

Here, you are talking about the sequence number. 

So. How would you like to go about it?

E.g. if you work on it in a fork, we can still be in discussion here and in a PR. 

Requirements that you do not have could be Python version compatibility and zoneinfo compatibility, performance.

For performance, I have a wondering about whether generating all events from 1970 to 2050 may or may not be so bad if at least they are skipped easily.
Also, your code could run optionally. We can make the tests so they switch it on and off.

What do you do with the RDATE if you do not consider EXDATE?

**Comment by niccokunzmann** [comment]
I modified the code to include "RECURRENCE-ID": https://github.com/niccokunzmann/python-recurring-ical-events/issues/90#issuecomment-1127200825

**Comment by NicoHood** [comment]
It took me a long time to understand your code, but it is super smart. The idea is to use the id + date OR recurring id in combination, which is mostly equivalent to my code, just smaller. What is not handled yet is the rdate, but that could be added as well. I have no sample whith an rdate, but isn't that just like the sequence?

This way we could generate new UIDs that are truely unique. Smart, smart...

Now the next step for me is to ask one more question:
Why are `RECURRENCE-ID` and such information even included in the generated event? Since the library's goal is to unfold all events I would expect them to not have this number.

In this particular usecase it is very useful to know, but if we integrate the logic inside the library, we can strip all the recurring related information. But this would also mean, that we need to overwrite the UID and set a new one.

In the end we have to decide:
* Do we want each outputted event to have a **unique** (new) id?
* Do we want to remove all (not required) information about the recurring, as the event was flattened?

I'd say yes, what do you say?

**Comment by NicoHood** [comment]
I have to correct my statement:
Using the date is not enough, as it would not work when the event was moved. Let's say you've added a recurring event on monday, but now it has moved from 19:00 to 20:00, then it would be treated as different UID. It would be better, if it still has the same uid, with the new date, so I can track that change and display it on my User Interface. That is the only option I have.

**Comment by niccokunzmann** [comment]
> Why are RECURRENCE-ID and such information even included in the generated event? Since the library's goal is to unfold all events I would expect them to not have this number.

This is partly due to the TDD aproach. Minimal effort. Also, which attributes are removed/altered is documented. No-one had a problem with having it there, so it is there sometimes. It could be added to every event that is recurring in order to identify them in a more consistent way or removed completely - since it is not specified now, we have a free hand without breaking neither our promise nor hopefully someone's expectations.

**Comment by NicoHood** [comment]
Well the recurrence id is usually set for moved events, but not for normal recurring events, right?

Indicating if an event is recurring is important to me. If we should use the recurrence id, i am not sure. It could be a smart default, very close to the spec. But it might also be very non-standard, I don't know.

**Comment by niccokunzmann** [comment]
From the standard RFC 5545 (referenced in README):

UID, RECURRENCE-ID and SEQUENCE are used to identify an individual instance within the recurrence set. - essence from screenshot

To me, an approach that is as wide as the standard, not narrower, not wider would use these three to identify events.

I do not know if it solves something for you. To me it seems essential - if we want to identify events - that these three properties are present in each of the events.

Also, depending on the software hosting the event calendar, they might not correctly use these three.

**Comment by NicoHood** [comment]
So I suggest to always set the recurrence id to the start of the event, if it IS a recurring event. This covers one of my usecases.

For the second usecase I am not yet sure what is the best practice. Rewriting the uid could be an idea.

**Comment by niccokunzmann** [comment]
If a simple thing can be done already, tests and PR are welcome.

What is your second use-case?

**Comment by NicoHood** [comment]
The second usecase is to have unique IDs for each event. None of the event should have the same UID. I'd suggest to use a combination of the mentioned UID + Sequence of the recurring event (note: the "sequence" I mean is not the sequence in the ical spec)

**Comment by niccokunzmann** [comment]
So - which information would you pass through to the list of events? Which attributes of the events?

**Comment by niccokunzmann** [comment]
My idea would be - if people make the changes to the events - that there is always some time between the changes. E.g. if an event is moved, one will be deleted and another pops up - if there is a minimal change in the other attributes like description, you can assume it is the same. Also, it may vary with the hosting software for event creation.

My approach would be TDD again - who would have thought: gather all different inputs and all different outputs. Maybe you do not want to work on the unfolded events but at the parsed icalendar and later when approved pass it to this library.
Your solution looks like something where you copy over events that need approval - nice one! If that is open-source, that is a great addition to the open web calendar eco-system.

**Comment by NicoHood** [comment]
I think we are misunderstanding each other. In the meantime I learned and tried a lot and got some good feeling of what I really need now.

I have two requirements for the **unfolded** events:
* Know if they were in a recurring series (yes/no)
* Know the number of recurrence with in the series (1...n)

I have implemented it by scanning the unfolded events afterwards, but it would make sense to integrate that into the library itself. We can keep everything as it is, but we need to add one more custom property. I think there is no way around a custom property, as this information is nothing specified by the spec.

I've implemented it with a single unsigned integer.
* If it is 0 the event is not recurring, if it is >0 it is (first requirement)
* A recurring event will have the number within that series stored as int starting from 1..n

I would keep the `UID` and I would keep the `RECURRENCE-ID` a) to not break existing code and b) as they are both useful in its own way. The UID is the original id and thereby useful (should be self explaining), the RECURRENCE-ID is the date from where the event date was *moved*. I personally do not (yet) need this information, but others might need.

So how do we want to name the new property? I've named it internally in my code `sequence`, but i think in the spec this name is already used for something else. `occurence` would be another idea, or `recurring-sequence`? Maybe also with a `X-` prefix?

Is it more clear now? If not, we can also have a chat in german, that might be easier.

**Comment by niccokunzmann** [comment]
I think, this is very clear now.

1. If an event is not recurring, nothing is changed
2. If an event is recurring, a property is added, I propose to name it `X-RECURRENCE-INDEX` as
    - it is about recurrence
    - it is a custom property
    - it is an index in a sequence.

I am happy to see a PR with some tests. I think these will make sure that you do not get any problems in production:

- [ ] no property
- [ ] what is expected if RDATE is there
- [ ] what is expected if only EXDATE is there
- [ ] what is expected if RRULE is there
- [ ] what happens if they are there but there is only one event in the sequence
- [ ] `between()` and `at()` will sometimes not return the event with index 0.
- [ ] create a test with about 10000 events (daily 1970 to 2000) and only use 1 of them
- [ ] proper index for rrule, exdate and rdate mix also with edited event

Did I get it? Hab ich es endlich verstanden? :)

By the way: if you like me to implement it, I am happy to be funded. I am doing this in my free time, have two kids and we live below the poverty line according to [this definition](https://research.senedd.wales/research-articles/poverty-in-wales-are-we-getting-the-full-picture/) (I am in Wales), at the moment.

**Comment by NicoHood** [comment]
That is correct so far. Just one difference: I've set the recurrence index to 0 for non recurring events. Just to indicate, that this feature is available (instead of leaving it out). Both options however should work for the usecase.

I need some more time to implement this, but we have a good starting point. Give me some time :)
